### PR TITLE
core: better logging when spawn() fails

### PR DESCRIPTION
**Pull request by tych0**
If someone sends us an invalid thing to run, right now we just silently
swallow it, because the logging module doesn't work after we've forked.

Instead, let's just keep around what we're going to exec (being smart about
shell=True invocations so we don't just do a lookup of "/bin/sh") and do a
path lookup on it. If it doesn't exist, we can fail without doing all the
forking and give the user a real error message.

While arguably not logging anything is fine (someone typed an invalid
thing), I've seen a lot of confusion more recently about spawn not working
because of typos and such. Hopefully this will fix things.

Signed-off-by: Tycho Andersen <tycho@tycho.pizza>

### CODE REVIEW

File: libqtile/core/manager.py [hunk 1]
```diff
@@ -1138,14 +1139,24 @@ def cmd_spawn(self, cmd, shell=False):
 
             spawn(["xterm", "-T", "Temporary terminal"])
         """
+        to_lookup = ""
         if shell:
             if not isinstance(cmd, str):
+                to_lookup = cmd[0]
                 cmd = subprocess.list2cmdline(cmd)
+            else:
+                to_lookup = cmd
```
Comment (ramnes) [inline libqtile/core/manager.py:1148]:
```suggestion
                to_lookup = cmd.split(" ", 1)[0]
```

File: libqtile/core/manager.py [hunk 2]
```diff
@@ -1138,14 +1139,24 @@ def cmd_spawn(self, cmd, shell=False):
...
             args = ["/bin/sh", "-c", cmd]
         elif isinstance(cmd, str):
             args = shlex.split(cmd)
+            to_lookup = args[0]
         else:
             args = list(cmd)
+            to_lookup = args[0]
```
Comment (ramnes) [inline libqtile/core/manager.py:1155]: maybe worth refactoring to look for the type of `cmd` before checking `shell`

File: libqtile/core/manager.py (hunk 1 above)
Comment (tych0) [reply libqtile/core/manager.py:1148]: Really it should be shlex.split(cmd)[0], right?

File: libqtile/core/manager.py (hunk 2 above)
Comment (tych0) [reply libqtile/core/manager.py:1155]: Yeah, that might be a bit better, I'll take a look.

File: libqtile/core/manager.py (hunk 1 above)
Comment (ramnes) [reply libqtile/core/manager.py:1148]: yeah preferably I guess, although it probably doesn't change anything because we're looking for the binary name

Comment (tych0) [reply libqtile/core/manager.py:1148]: It does if the binary name has a space in it :)

File: libqtile/core/manager.py (hunk 2 above)
Comment (tych0) [reply libqtile/core/manager.py:1155]: So, I couldn't figure out how to do this nicely. Turns out that if `shell=True`, we need to render `cmd` as a string, but if `shell=False`, we want it as a list. I'm open to other implementations, though.

Comment (ramnes) [reply libqtile/core/manager.py:1155]: This is the best I could come with:

```python
if isinstance(cmd, str):
    args = shlex.split(cmd)
    cmd = subprocess.list2cmdline(args)
else:
    args = list(cmd)

to_lookup = args[0]
if shell:
    args = ["/bin/sh", "-c", cmd]
```

Comment (tych0) [reply libqtile/core/manager.py]: Cool, thanks, I updated it to this.

**Pull request merged**